File: server/modules/workspace/path-guard.ts
import { resolve, relative, isAbsolute } from "node:path";

export function assertPathWithinRoot(root: string, relativePath: string): string {
  const resolvedRoot = resolve(root);
  const resolvedPath = resolve(resolvedRoot, relativePath);
  const relativeToRoot = relative(resolvedRoot, resolvedPath);

  if (
    relativeToRoot === "" ||
    (!relativeToRoot.startsWith("..") && !isAbsolute(relativeToRoot))
  ) {
    return resolvedPath;
  }

  throw new Error(
    `Path escapes worktree: "${relativePath}" resolves outside "${resolvedRoot}"`,
  );
}

File: server-go/internal/workspace/path_guard.go
package workspace

import (
	"fmt"
	"path/filepath"
	"strings"
)

func AssertPathWithinRoot(root, relativePath string) (string, error) {
	resolvedRoot, err := filepath.Abs(root)
	if err != nil {
		return "", err
	}
	resolvedPath, err := filepath.Abs(filepath.Join(resolvedRoot, relativePath))
	if err != nil {
		return "", err
	}
	rel, err := filepath.Rel(resolvedRoot, resolvedPath)
	if err != nil {
		return "", err
	}
	if rel == "" || (!strings.HasPrefix(rel, "..") && !filepath.IsAbs(rel)) {
		return resolvedPath, nil
	}
	return "", fmt.Errorf("path escapes worktree: %q resolves outside %q", relativePath, resolvedRoot)
}
